### PR TITLE
[Graph] Mutation for Heterograph

### PR DESCRIPTION
**Pull request by classicsong**
## Description
- [x] add_nodes
- [x] add_edges
- [x] remove_nodes
- [x] remove_edges
- [x] add_selfloop
- [x] remove_selfloop

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/convert.py [hunk 1]
```diff
@@ -501,8 +501,10 @@ def heterograph(data_dict, num_nodes_dict=None, restrict_format='auto',
         num_nodes_dict = defaultdict(int)
         for (srctype, etype, dsttype), data in data_dict.items():
             if isinstance(data, tuple):
-                src = utils.toindex(data[0]).tonumpy()
-                dst = utils.toindex(data[1]).tonumpy()
+                src = utils.toindex(data[0],
+                                    dtype='int32' if idtype == F.int32 else 'int64').tonumpy()
+                dst = utils.toindex(data[1],
+                                    dtype='int32' if idtype == F.int32 else 'int64').tonumpy()
```
Comment (jermainewang) [inline python/dgl/convert.py:507]: Add a TODO here:
`TODO(minjie): converting to index is unnecessary just for getting the max ID.`
I will change them later when I get rid of all the Index object.

File: python/dgl/heterograph.py [hunk 2]
```diff
@@ -311,28 +311,618 @@ def __repr__(self):
...
-        raise DGLError('Mutation is not supported in heterograph.')
+        # TODO(xiangsx): block do not support add_nodes
+        if ntype is None:
+            if self._graph.number_of_ntypes() != 1:
+                raise DGLError('Node type name must be specified if there are more than one '
+                               'node types.')
+
+        # nothing happen
+        if num == 0:
+            return self
```
Comment (jermainewang) [inline python/dgl/heterograph.py:407]: `return self` -> `return`

File: python/dgl/heterograph.py [hunk 3]
```diff
@@ -311,28 +311,618 @@ def __repr__(self):
...
+            else:
+                nodes[c_ntype] = self.nodes(c_ntype)
+
+        # node_subgraph
+        sub_g = self.subgraph(nodes)
+        self._graph = sub_g._graph
+        self._node_frames = sub_g._node_frames
+        self._edge_frames = sub_g._edge_frames
+
+    def add_selfloop(self):
```
Comment (jermainewang) [inline python/dgl/heterograph.py:806]: For `add_selfloop` and `remove_selfloop`, how about just adding an `etype` argument and letting users select which relation to modify? We raise error for edge type that does not that represent a homograph.

Comment (classicsong) [reply python/dgl/heterograph.py]: Done

File: python/dgl/convert.py (hunk 1 above)
Comment (classicsong) [reply python/dgl/convert.py:508]: Done

File: python/dgl/heterograph.py (hunk 2 above)
Comment (classicsong) [reply python/dgl/heterograph.py]: Done.

**Pull request merged**